ocaml/xapi-storage-script/main.ml

```diff
@@ -1334,6 +1334,31 @@ let bind ~volume_script_dir =
     |> wrap
   in
   S.VDI.activate3 vdi_activate3_impl ;
+  let vdi_activate_readonly_impl dbg _dp sr vdi' vm' =
+    (let vdi = Storage_interface.Vdi.string_of vdi' in
+     let domain = Storage_interface.Vm.string_of vm' in
+     Attached_SRs.find sr >>>= fun sr ->
+     (* Discover the URIs using Volume.stat *)
+     stat ~dbg ~sr ~vdi >>>= fun response ->
+     (* If we have a clone-on-boot volume then use that instead *)
```

Comment:
Can't you reuse 'vdi_attach_common' for this part? The only thing that changes i the Datapath_client.<fun> to be called, so vdi_attach_common could be changed to take an ~f to invoke

Reply:
Unfortunately, the return types of `Datapath_client.attach` and `Datapath_client.activate` are different so that doesn't quite work out. I could probably still share code between the two activate calls.